File: RI_sample.py
#/usr/bin/python
# Sample source code for "Recurrent Infomax generates synfire chains,
# neuronal avalanches, and simple cell-like selectivity"
# (Neural Computation 21(4) 1038-1067) by Takuma Tanaka (tanaka.takuma@gmail.com).
# To reproduce Fig. 6
# $ python RI_sample.py -N 50 --pbar 0.05 --pmax 0.95 -B 2500 -S 0 -T 50000 --name sequence
#
# To reproduce Fig. 8
# $ python RI_sample.py -N 50 --pbar 0.01 --pmax 0.5 -B 20000 -S 5000 -T 10000 --name avalanche

import os;
import sys;
import math;
import numpy;
import random;
import argparse;

parser = argparse.ArgumentParser(description="Recurrent Infomax sample program");
parser.add_argument("-N", action="store", type=int, dest="N", default=20, help="number of neurons");
parser.add_argument("--pbar", action="store", type=float, dest="pbar", default=0.01, help="mean firing rate");
parser.add_argument("--pmax", action="store", type=float, dest="pmax", default=0.95, help="reliability of the neurons");
parser.add_argument("-B", action="store", type=int, dest="B", default=10000, help="number of learning blocks");
parser.add_argument("-S", action="store", type=int, dest="S", default=0, help="number of blocks counting the number of avalanches");
parser.add_argument("-T", action="store", type=int, dest="T", default=10000, help="steps in a block");
parser.add_argument("--epsilon", action="store", type=float, dest="epsilon", default=0.01, help="learning rate of h");
parser.add_argument("--eta", action="store", type=float, dest="eta", default=0.2, help="learning rate of W");
parser.add_argument("--initialskip", action="store", type=int, dest="initialskip", default=5, help="initial blocks without the update of W");
parser.add_argument("--loginterval", action="store", type=int, dest="loginterval", default=5, help="logging interval");
parser.add_argument("--name", action="store", type=str, dest="name", required=True, help="output file name");
parser.add_argument("--seed", action="store", type=int, dest="seed", default=1, help="seed of random number generators");

args = parser.parse_args();
N = args.N;
pbar = args.pbar;
pmax = args.pmax;
B = args.B;
S = args.S;
T = args.T;
loginterval = args.loginterval;
initialskip = args.initialskip;
epsilon = args.epsilon;
eta = args.eta;
name = args.name;
seed = args.seed;

random.seed(seed);
numpy.random.seed(seed);
meanI = 0;

x = numpy.array([0]*N);
W = numpy.random.rand(N,N)-0.5;
h = numpy.array([0.]*N);
p = numpy.array([pbar]*N);

histmax = 1000;
shist = [0]*histmax;

myfile = open("log_"+name+".txt", "w");

burstsize = 0;
raster = [0]*T;

def onestep(corrmode):
    global x, h, W, burstsize, shist, C, Eyx;
    y = (numpy.sign(pmax/(1+numpy.exp(-W.dot(x)+h))-numpy.random.rand(N))+1)/2;
    h += epsilon*(y-p);
    if corrmode:
        C += numpy.outer(y-p, y-p);
        Eyx += numpy.outer(y-p, x-p);
    x = y;
    if longiter>=B-S:
        if burstsize>0:
            if numpy.sum(x)==0:
                shist[burstsize] += 1;
                burstsize = 0;
            else:
                burstsize += int(numpy.sum(x));
        elif numpy.sum(x)>0:
            burstsize = int(numpy.sum(x));

for longiter in range(B):
    for shortiter in range(2*T):
        onestep(False);

    C = numpy.zeros((N,N));
    Eyx = numpy.zeros((N,N));
    for shortiter in range(T):
        onestep(True);
        if longiter==B-1:
            raster[shortiter] = x;
    C /= T;
    Eyx /= T;
    
    D = numpy.r_[numpy.c_[C, numpy.transpose(Eyx)], numpy.c_[Eyx, C]];
    C_ = numpy.linalg.inv(C);
    CC_ = numpy.r_[numpy.c_[C_, numpy.zeros((N, N))], numpy.c_[numpy.zeros((N, N)), C_]];
    D_ = numpy.linalg.inv(CC_.dot(D)).dot(CC_);
    if initialskip<=longiter and longiter<B-S:
        V = (numpy.ones((N, N))-numpy.eye(N))*(2*C_-D_[0:N,0:N]-D_[N:2*N,N:2*N]);
        D_t = numpy.transpose(D_);
        U = (numpy.outer(1-2*p, 1-2*p)*Eyx+numpy.outer(p*(1-p), p*(1-p))-Eyx*Eyx)*(D_[N:2*N,0:N]+D_t[N:2*N,0:N]);
        W += eta*(numpy.transpose(C).dot(V).dot(Eyx)+numpy.transpose(numpy.transpose(Eyx).dot(V).dot(C))-U)/2;
    meanI += -math.log(numpy.linalg.det(CC_.dot(D)))/(2*math.log(2.0));
    
    if longiter%loginterval==loginterval-1:
        message = str(longiter+1)+" "+str(meanI/loginterval)+"\n";
        sys.stdout.write(message)
        myfile.write(message)
        myfile.flush();
        meanI = 0;
myfile.close();

if S>0:
    myfile = open("hist_avalanche_"+name+".txt", "w");
    for i in range(histmax):
        if shist[i]>0:
            myfile.write(str(i)+" "+str(shist[i]/float(sum(shist)))+"\n");
    myfile.close();

braster = [0]*T;
myfile = open("raster_"+name+".txt", "w");
for i in range(T):
    bvec = "";
    for j in range(N):
        if raster[i][j]!=0:
            myfile.write(str(i)+" "+str(j)+"\n");
            bvec += "1";
        else:
            bvec += "0";
    braster[i] = bvec;
myfile.close();

repcount = [(x, braster.count(x)) for x in set(braster)];
rephist = [0]*N;
for x,num in repcount:
    rephist[x.count('1')] += num;
myfile = open("pattern_repetition_"+name+".txt", "w");
for i in range(1,N):
    if rephist[i]>0:
        myfile.write(str(i)+" "+str(rephist[i])+"\n");
myfile.close();

myfile = open("connection_"+name+".txt", "w");
for i in range(N):
    for j in range(N):
        myfile.write(str(W[i][j])+" ");
    myfile.write("\n");
myfile.close();

File: RI_sample.m
function RI_sample(mode)

% Sample source code for "Recurrent Infomax generates synfire chains,
% neuronal avalanches, and simple cell-like selectivity"
% (Neural Computation 21(4) 1038-1067) by Takuma Tanaka.
%
% This source code runs on Matlab and Octave.
% If you find any bugs in this program, or have any question
% I would appreciate the tip
% (ttakuma@mbs.med.kyoto-u.ac.jp).
%
% To run this program:
%
% 1. If you are using Matlab, delete the line of code 'fflush(stdout);' 
% appearing later on in this file.
%
% 2. Go into the directory where the file RI_sample.m is located, and enter
%      matlab
% or
%      octave
% in the console.
%
% 3. Enter
%      RI_sample(1)
% for avalanche before learning (Fig. 8, N=50, p_max=0.5, \bar{p}_i=0.01),
%      RI_sample(2)
% for avalanche after learning (Fig. 8, N=50, p_max=0.5, \bar{p}_i=0.01),
%      RI_sample(3)
% for sequences before learning (Fig. 6, N=50, p_max=0.95, \bar{p}_i=0.05), or
%      RI_sample(4)
% for sequences after learning (Fig. 6, N=50, p_max=0.95, \bar{p}_i=0.05),
% and then the simulation will run.
%
% The simulation takes several hours to days in modes 2 and 4.
%
% The block number and approximate mutual information are displayed in
% the console and also output to the log files named 
% 'log_avalanche_before.txt' etc.
% At the end of the simulation, the histogram of the burst size is output to
% the files 'hist_avalanche_before.txt' and 'hist_avalanche_after.txt'
% in modes 1 and 2, respectively.
% Firing states of the neurons in the last block are output to the files
% 'raster_sequence_before.txt' and 'raster_sequence_after.txt'
% in modes 3 and 4, respectively.
% The numbers of repeated patterns are output to the files
% 'number_sequence_before.txt' and 'number_sequence_after.txt'
% in modes 3 and 4, respectively.
% The sequence size is output in the first column, and the number of
% occurrences is output in the second column.
% Optimized connection weights are output to the files named
% 'connection_avalanche_before.txt' etc.

if (mode~=1 & mode~=2 & mode~=3 & mode~=4)
  fprintf(1, 'Invalid mode.\n  Mode 1: avalanche before learning\n  Mode 2: avalanche after learning\n  Mode 3: sequence before learning\n  Mode 4: sequence after learning\n');
  return;
end;

rand('seed', 1);
meanI = 0;
logstep = 5;
factor = 100;

N = 50;
x = zeros(N, 1);
W = rand(N,N)-0.5;
h = zeros(N, 1);
epsilon = 0.01;
eta = 0.2;

histmax = 1000;
shist = zeros(histmax, 1);

if (mode==1 | mode==2)
  p = 0.01*ones(N, 1);
  pmax = 0.5;
  T = 10000;
  if (mode==1)
    fname = 'log_avalanche_before.txt';
    STAT = 5;
    BLOCKS = 5005;
  else
    fname = 'log_avalanche_after.txt';
    STAT = 15000;
    BLOCKS = 20000;
  end
else
  STAT = 5000;
  p = 0.05*ones(N, 1);
  pmax = 0.95;
  T = 50000;
  if (mode==3)
    fname = 'log_sequence_before.txt';
    BLOCKS = 5;
  else
    fname = 'log_sequence_after.txt';
    BLOCKS = 2500;
  end;
  raster = zeros(T, N);
end

myfile = fopen(fname, 'w', 'ieee-le');

state = 0;

for longiter=1:BLOCKS
  for shortiter=1:T
    y = pmax ./ (1 + exp(-W*x+h));
    y = rand(N, 1)<y;
    h = h+epsilon*(y-p);
    x = y;
    
    if (longiter>=STAT)
      if (state>0)
        if (sum(x)==0)
          shist(state) = shist(state)+1;
          state = 0;
        else
          state = state+sum(x);
        end
      elseif (sum(x)>0)
        state = sum(x);
      end;
    end;
  end;

  C = zeros(N,N);
  Eyx = zeros(N,N);

  for shortiter=1:T
    y = pmax ./ (1 + exp(-W*x+h));
    y = rand(N, 1)<y;
    h = h+epsilon*(y-p);
    C = C+(y-p)*(y-p)';
    Eyx = Eyx+(y-p)*(x-p)';
    x = y;
    if (longiter>=STAT)
      if (state>0)
        if (sum(x)==0)
          shist(state) = shist(state)+1;
          state = 0;
        else
          state = state+sum(x);
        end
      elseif (sum(x)>0)
        state = sum(x);
      end;
    elseif (longiter==BLOCKS & mode>=3)
      raster(shortiter,:) = x';
    end;
  end;

  C = C/T;
  Eyx = Eyx/T;
  D = [C Eyx' ; Eyx C];
  
  if (5<longiter & longiter<STAT)
    C_ = (factor*C)'^-1;
    CC_ = [C_ zeros(N) ; zeros(N) C_];
    D_ = ((factor*D*CC_)'^(-1))*CC_;
    V = (ones(N, N)-eye(N)) .* (2*C_-D_(1:N,1:N)-D_(N+1:2*N,N+1:2*N))*factor;
    D_t = D_';
    U = (((1-2*p)*(1-2*p)').*Eyx+(p.*(1-p))*(p.*(1-p))'-Eyx.*Eyx).*((D_(N+1:2*N,1:N))+(D_t(N+1:2*N,1:N)))*factor;
    W = W+eta*(C'*V*Eyx+(Eyx'*V*C)'-U)/2;
  end;

  I = (log(det(factor*C))-log(det(factor*D))/2)/log(2);
  meanI = meanI+I;
  
  fprintf(1, '%d ', longiter);
  disp(meanI/(mod(longiter-1, logstep)+1));

  fflush(stdout);  % Delete this line if you are using Matlab
  
  if (mod(longiter, logstep)==0)
    fprintf(myfile, '%d %e\n', longiter, meanI/logstep);
    if (mod(longiter, 100)==0)
      fclose(myfile);
      myfile = fopen(fname, 'a', 'ieee-le');
    end;
    meanI = 0;
  end;
  
end;

fclose(myfile);

if (mode<=2)
  if (mode==1)
    myfile2 = fopen('hist_avalanche_before.txt', 'w', 'ieee-le');
  else
    myfile2 = fopen('hist_avalanche_after.txt', 'w', 'ieee-le');
  end
  for i=1:histmax
    if (shist(i)>0)
      fprintf(myfile2, '%d %e\n', i, shist(i)/sum(shist));
    end;
  end;
else
  raster = [raster zeros(T, 2)];
  if (mode==3)
    myfile2 = fopen('raster_sequence_before.txt', 'w', 'ieee-le');
  else
    myfile2 = fopen('raster_sequence_after.txt', 'w', 'ieee-le');
  end;
  for i=1:T
    index = 0;
    for j=1:N
      fprintf(myfile2, '%d ', raster(i,j));
      index = index*2+raster(i,j);
    end;
    fprintf(myfile2, '\n');
    raster(i, N+1) = sum(raster(i, :));
    raster(i, N+2) = index;
  end;
  fclose(myfile2);
  [s, perm] = sort(raster(:,N+2));
  raster = raster(perm, :);
  [s, perm] = sort(raster(:,N+1));
  raster = raster(perm, :);

  repeated = zeros(N, 1);
  if (raster(1,N+2)==raster(2,N+2) & raster(2,N+1)>0)
    repeated(raster(1, N+1), 1) = repeated(raster(1, N+1), 1)+1;
  end;
  for i=2:T-1
    if (raster(i,N+1)>0 & (raster(i,N+2)==raster(i-1,N+2) | raster(i,N+2)==raster(i+1,N+2)))
      repeated(raster(i, N+1), 1) = repeated(raster(i, N+1), 1)+1;
    end;
  end;
  if (raster(T,N+1)>0 & raster(T,N+2)==raster(T-1,N+2))
    repeated(raster(T, N+1), 1) = repeated(raster(T, N+1), 1)+1;
  end;

  if (mode==3)
    myfile2 = fopen('number_sequence_before.txt', 'w', 'ieee-le');
  else
    myfile2 = fopen('number_sequence_after.txt', 'w', 'ieee-le');
  end;
  for i=1:N
    if (repeated(i)>0)
      fprintf(myfile2, '%d %d\n', i, repeated(i));
    end;
  end;
end;

fclose(myfile2);

if (mode==1)
  myfile3 = fopen('connection_avalanche_before.txt', 'w', 'ieee-le');
elseif (mode==2)
  myfile3 = fopen('connection_avalanche_after.txt', 'w', 'ieee-le');
elseif (mode==3)
  myfile3 = fopen('connection_sequence_before.txt', 'w', 'ieee-le');
else
  myfile3 = fopen('connection_sequence_after.txt', 'w', 'ieee-le');
end;

for i=1:N
  for j=1:N
    fprintf(myfile3, '%e ', W(i,j));
  end;
  fprintf(myfile3, '\n');
end;

fclose(myfile3);

